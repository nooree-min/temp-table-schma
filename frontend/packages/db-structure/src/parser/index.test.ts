// biome-ignore lint/correctness/noNodejsModules: this file is only used in tests
import fs from 'node:fs'
// biome-ignore lint/correctness/noNodejsModules: this file is only used in tests
import path from 'node:path'
import { describe, expect, it } from 'vitest'
import { parse } from './index.js'

describe(parse, () => {
  it('should parse schema.rb to JSON correctly', async () => {
    const schemaText = fs.readFileSync(
      path.resolve(__dirname, './schemarb/input/schema1.in.rb'),
      'utf-8',
    )

    const { value } = await parse(schemaText, 'schemarb')
    expect(value).toMatchSnapshot()
  })

  it('should parse postgresql to JSON correctly', async () => {
    const schemaText = fs.readFileSync(
      path.resolve(__dirname, './sql/input/postgresql_schema1.in.sql'),
      'utf-8',
    )

    const { value } = await parse(schemaText, 'postgres')
    expect(value).toMatchSnapshot()
  })
})
