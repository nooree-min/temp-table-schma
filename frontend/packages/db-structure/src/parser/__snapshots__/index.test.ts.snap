// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse > should parse postgresql to JSON correctly 1`] = `
{
  "relationships": {
    "users_id_to_posts_user_id": {
      "cardinality": "ONE_TO_MANY",
      "deleteConstraint": "NO_ACTION",
      "foreignColumnName": "user_id",
      "foreignTableName": "posts",
      "name": "users_id_to_posts_user_id",
      "primaryColumnName": "id",
      "primaryTableName": "users",
      "updateConstraint": "NO_ACTION",
    },
  },
  "tableGroups": {},
  "tables": {
    "posts": {
      "columns": {
        "id": {
          "check": null,
          "comment": null,
          "default": null,
          "name": "id",
          "notNull": true,
          "primary": true,
          "type": "serial",
          "unique": true,
        },
        "user_id": {
          "check": null,
          "comment": null,
          "default": null,
          "name": "user_id",
          "notNull": false,
          "primary": false,
          "type": "int4",
          "unique": false,
        },
      },
      "comment": null,
      "constraints": {},
      "indexes": {},
      "name": "posts",
    },
    "users": {
      "columns": {
        "created_at": {
          "check": null,
          "comment": null,
          "default": null,
          "name": "created_at",
          "notNull": false,
          "primary": false,
          "type": "timestamp",
          "unique": false,
        },
        "email": {
          "check": null,
          "comment": null,
          "default": null,
          "name": "email",
          "notNull": true,
          "primary": false,
          "type": "varchar",
          "unique": true,
        },
        "id": {
          "check": null,
          "comment": null,
          "default": null,
          "name": "id",
          "notNull": true,
          "primary": true,
          "type": "serial",
          "unique": true,
        },
        "username": {
          "check": null,
          "comment": "user fullname.",
          "default": null,
          "name": "username",
          "notNull": true,
          "primary": false,
          "type": "varchar",
          "unique": true,
        },
      },
      "comment": "store our users.",
      "constraints": {},
      "indexes": {
        "index_users_on_id_and_email": {
          "columns": [
            "id",
            "email",
          ],
          "name": "index_users_on_id_and_email",
          "type": "btree",
          "unique": true,
        },
      },
      "name": "users",
    },
  },
}
`;

exports[`parse > should parse schema.rb to JSON correctly 1`] = `
{
  "relationships": {
    "fk_comments_task_id": {
      "cardinality": "ONE_TO_MANY",
      "deleteConstraint": "CASCADE",
      "foreignColumnName": "task_id",
      "foreignTableName": "comments",
      "name": "fk_comments_task_id",
      "primaryColumnName": "id",
      "primaryTableName": "tasks",
      "updateConstraint": "RESTRICT",
    },
    "fk_comments_user_id": {
      "cardinality": "ONE_TO_MANY",
      "deleteConstraint": "CASCADE",
      "foreignColumnName": "user_id",
      "foreignTableName": "comments",
      "name": "fk_comments_user_id",
      "primaryColumnName": "id",
      "primaryTableName": "users",
      "updateConstraint": "RESTRICT",
    },
    "fk_departments_company_id": {
      "cardinality": "ONE_TO_MANY",
      "deleteConstraint": "CASCADE",
      "foreignColumnName": "company_id",
      "foreignTableName": "departments",
      "name": "fk_departments_company_id",
      "primaryColumnName": "id",
      "primaryTableName": "companies",
      "updateConstraint": "RESTRICT",
    },
    "fk_project_assignments_project_id": {
      "cardinality": "ONE_TO_MANY",
      "deleteConstraint": "CASCADE",
      "foreignColumnName": "project_id",
      "foreignTableName": "project_assignments",
      "name": "fk_project_assignments_project_id",
      "primaryColumnName": "id",
      "primaryTableName": "projects",
      "updateConstraint": "RESTRICT",
    },
    "fk_project_assignments_user_id": {
      "cardinality": "ONE_TO_MANY",
      "deleteConstraint": "CASCADE",
      "foreignColumnName": "user_id",
      "foreignTableName": "project_assignments",
      "name": "fk_project_assignments_user_id",
      "primaryColumnName": "id",
      "primaryTableName": "users",
      "updateConstraint": "RESTRICT",
    },
    "fk_projects_company_id": {
      "cardinality": "ONE_TO_MANY",
      "deleteConstraint": "CASCADE",
      "foreignColumnName": "company_id",
      "foreignTableName": "projects",
      "name": "fk_projects_company_id",
      "primaryColumnName": "id",
      "primaryTableName": "companies",
      "updateConstraint": "RESTRICT",
    },
    "fk_tasks_assigned_user_id": {
      "cardinality": "ONE_TO_MANY",
      "deleteConstraint": "RESTRICT",
      "foreignColumnName": "assigned_user_id",
      "foreignTableName": "tasks",
      "name": "fk_tasks_assigned_user_id",
      "primaryColumnName": "id",
      "primaryTableName": "users",
      "updateConstraint": "RESTRICT",
    },
    "fk_tasks_project_id": {
      "cardinality": "ONE_TO_MANY",
      "deleteConstraint": "CASCADE",
      "foreignColumnName": "project_id",
      "foreignTableName": "tasks",
      "name": "fk_tasks_project_id",
      "primaryColumnName": "id",
      "primaryTableName": "projects",
      "updateConstraint": "RESTRICT",
    },
    "fk_timesheets_task_id": {
      "cardinality": "ONE_TO_MANY",
      "deleteConstraint": "CASCADE",
      "foreignColumnName": "task_id",
      "foreignTableName": "timesheets",
      "name": "fk_timesheets_task_id",
      "primaryColumnName": "id",
      "primaryTableName": "tasks",
      "updateConstraint": "RESTRICT",
    },
    "fk_timesheets_user_id": {
      "cardinality": "ONE_TO_MANY",
      "deleteConstraint": "CASCADE",
      "foreignColumnName": "user_id",
      "foreignTableName": "timesheets",
      "name": "fk_timesheets_user_id",
      "primaryColumnName": "id",
      "primaryTableName": "users",
      "updateConstraint": "RESTRICT",
    },
    "fk_user_roles_role_id": {
      "cardinality": "ONE_TO_MANY",
      "deleteConstraint": "CASCADE",
      "foreignColumnName": "role_id",
      "foreignTableName": "user_roles",
      "name": "fk_user_roles_role_id",
      "primaryColumnName": "id",
      "primaryTableName": "roles",
      "updateConstraint": "RESTRICT",
    },
    "fk_user_roles_user_id": {
      "cardinality": "ONE_TO_MANY",
      "deleteConstraint": "CASCADE",
      "foreignColumnName": "user_id",
      "foreignTableName": "user_roles",
      "name": "fk_user_roles_user_id",
      "primaryColumnName": "id",
      "primaryTableName": "users",
      "updateConstraint": "RESTRICT",
    },
    "fk_users_company_id": {
      "cardinality": "ONE_TO_MANY",
      "deleteConstraint": "CASCADE",
      "foreignColumnName": "company_id",
      "foreignTableName": "users",
      "name": "fk_users_company_id",
      "primaryColumnName": "id",
      "primaryTableName": "companies",
      "updateConstraint": "RESTRICT",
    },
    "fk_users_department_id": {
      "cardinality": "ONE_TO_MANY",
      "deleteConstraint": "CASCADE",
      "foreignColumnName": "department_id",
      "foreignTableName": "users",
      "name": "fk_users_department_id",
      "primaryColumnName": "id",
      "primaryTableName": "departments",
      "updateConstraint": "RESTRICT",
    },
  },
  "tableGroups": {},
  "tables": {
    "comments": {
      "columns": {
        "content": {
          "check": null,
          "comment": "The content of the comment.",
          "default": null,
          "name": "content",
          "notNull": true,
          "primary": false,
          "type": "text",
          "unique": false,
        },
        "created_at": {
          "check": null,
          "comment": "The timestamp when the comment was created.",
          "default": null,
          "name": "created_at",
          "notNull": true,
          "primary": false,
          "type": "timestamp",
          "unique": false,
        },
        "id": {
          "check": null,
          "comment": null,
          "default": null,
          "name": "id",
          "notNull": true,
          "primary": true,
          "type": "uuid",
          "unique": true,
        },
        "task_id": {
          "check": null,
          "comment": "Foreign key linking the comment to a task.",
          "default": null,
          "name": "task_id",
          "notNull": true,
          "primary": false,
          "type": "uuid",
          "unique": false,
        },
        "user_id": {
          "check": null,
          "comment": "Foreign key linking the comment to the user who wrote it.",
          "default": null,
          "name": "user_id",
          "notNull": true,
          "primary": false,
          "type": "uuid",
          "unique": false,
        },
      },
      "comment": "Stores comments on tasks, enabling discussions or updates.",
      "constraints": {},
      "indexes": {
        "index_comments_on_task_id": {
          "columns": [
            "task_id",
          ],
          "name": "index_comments_on_task_id",
          "type": "",
          "unique": false,
        },
        "index_comments_on_user_id": {
          "columns": [
            "user_id",
          ],
          "name": "index_comments_on_user_id",
          "type": "",
          "unique": false,
        },
      },
      "name": "comments",
    },
    "companies": {
      "columns": {
        "address": {
          "check": null,
          "comment": "The registered address of the company.",
          "default": null,
          "name": "address",
          "notNull": true,
          "primary": false,
          "type": "varchar",
          "unique": false,
        },
        "id": {
          "check": null,
          "comment": null,
          "default": null,
          "name": "id",
          "notNull": true,
          "primary": true,
          "type": "uuid",
          "unique": true,
        },
        "name": {
          "check": null,
          "comment": "The name of the company.",
          "default": null,
          "name": "name",
          "notNull": true,
          "primary": false,
          "type": "varchar",
          "unique": false,
        },
      },
      "comment": "Represents organizations using the system. Each company is a top-level entity that owns departments, users, and projects.",
      "constraints": {},
      "indexes": {},
      "name": "companies",
    },
    "departments": {
      "columns": {
        "company_id": {
          "check": null,
          "comment": "Foreign key linking the department to a company.",
          "default": null,
          "name": "company_id",
          "notNull": true,
          "primary": false,
          "type": "uuid",
          "unique": false,
        },
        "id": {
          "check": null,
          "comment": null,
          "default": null,
          "name": "id",
          "notNull": true,
          "primary": true,
          "type": "uuid",
          "unique": true,
        },
        "name": {
          "check": null,
          "comment": "The name of the department.",
          "default": null,
          "name": "name",
          "notNull": true,
          "primary": false,
          "type": "varchar",
          "unique": false,
        },
      },
      "comment": "Represents departments within a company, organizing users into functional groups.",
      "constraints": {},
      "indexes": {
        "index_departments_on_company_id": {
          "columns": [
            "company_id",
          ],
          "name": "index_departments_on_company_id",
          "type": "",
          "unique": false,
        },
      },
      "name": "departments",
    },
    "project_assignments": {
      "columns": {
        "id": {
          "check": null,
          "comment": null,
          "default": null,
          "name": "id",
          "notNull": true,
          "primary": true,
          "type": "uuid",
          "unique": true,
        },
        "project_id": {
          "check": null,
          "comment": "Foreign key linking to the project.",
          "default": null,
          "name": "project_id",
          "notNull": true,
          "primary": false,
          "type": "uuid",
          "unique": false,
        },
        "user_id": {
          "check": null,
          "comment": "Foreign key linking to the user.",
          "default": null,
          "name": "user_id",
          "notNull": true,
          "primary": false,
          "type": "uuid",
          "unique": false,
        },
      },
      "comment": "Associates users with projects they are assigned to work on.",
      "constraints": {},
      "indexes": {
        "index_project_assignments_on_user_id_and_project_id": {
          "columns": [
            "user_id",
            "project_id",
          ],
          "name": "index_project_assignments_on_user_id_and_project_id",
          "type": "",
          "unique": true,
        },
      },
      "name": "project_assignments",
    },
    "projects": {
      "columns": {
        "company_id": {
          "check": null,
          "comment": "Foreign key linking the project to a company.",
          "default": null,
          "name": "company_id",
          "notNull": true,
          "primary": false,
          "type": "uuid",
          "unique": false,
        },
        "description": {
          "check": null,
          "comment": "A detailed description of the project.",
          "default": null,
          "name": "description",
          "notNull": false,
          "primary": false,
          "type": "text",
          "unique": false,
        },
        "id": {
          "check": null,
          "comment": null,
          "default": null,
          "name": "id",
          "notNull": true,
          "primary": true,
          "type": "uuid",
          "unique": true,
        },
        "name": {
          "check": null,
          "comment": "The name of the project.",
          "default": null,
          "name": "name",
          "notNull": true,
          "primary": false,
          "type": "varchar",
          "unique": false,
        },
      },
      "comment": "Represents projects managed within a company. Projects are linked to tasks and users.",
      "constraints": {},
      "indexes": {
        "index_projects_on_company_id": {
          "columns": [
            "company_id",
          ],
          "name": "index_projects_on_company_id",
          "type": "",
          "unique": false,
        },
      },
      "name": "projects",
    },
    "roles": {
      "columns": {
        "description": {
          "check": null,
          "comment": "A brief description of the role's purpose or permissions.",
          "default": null,
          "name": "description",
          "notNull": false,
          "primary": false,
          "type": "varchar",
          "unique": false,
        },
        "id": {
          "check": null,
          "comment": null,
          "default": null,
          "name": "id",
          "notNull": true,
          "primary": true,
          "type": "uuid",
          "unique": true,
        },
        "name": {
          "check": null,
          "comment": "The name of the role.",
          "default": null,
          "name": "name",
          "notNull": true,
          "primary": false,
          "type": "varchar",
          "unique": false,
        },
      },
      "comment": "Defines roles that can be assigned to users, such as 'Admin' or 'Manager'.",
      "constraints": {},
      "indexes": {},
      "name": "roles",
    },
    "tasks": {
      "columns": {
        "assigned_user_id": {
          "check": null,
          "comment": "Foreign key linking the task to the assigned user.",
          "default": null,
          "name": "assigned_user_id",
          "notNull": true,
          "primary": false,
          "type": "uuid",
          "unique": false,
        },
        "description": {
          "check": null,
          "comment": "A detailed description of the task.",
          "default": null,
          "name": "description",
          "notNull": false,
          "primary": false,
          "type": "text",
          "unique": false,
        },
        "due_date": {
          "check": null,
          "comment": "The deadline for completing the task.",
          "default": null,
          "name": "due_date",
          "notNull": false,
          "primary": false,
          "type": "timestamp",
          "unique": false,
        },
        "id": {
          "check": null,
          "comment": null,
          "default": null,
          "name": "id",
          "notNull": true,
          "primary": true,
          "type": "uuid",
          "unique": true,
        },
        "project_id": {
          "check": null,
          "comment": "Foreign key linking the task to a project.",
          "default": null,
          "name": "project_id",
          "notNull": true,
          "primary": false,
          "type": "uuid",
          "unique": false,
        },
        "status": {
          "check": null,
          "comment": "The current status of the task (e.g., 0: pending, 1: in progress, 2: completed).",
          "default": 0,
          "name": "status",
          "notNull": false,
          "primary": false,
          "type": "integer",
          "unique": false,
        },
        "title": {
          "check": null,
          "comment": "The title of the task.",
          "default": null,
          "name": "title",
          "notNull": true,
          "primary": false,
          "type": "varchar",
          "unique": false,
        },
      },
      "comment": "Represents tasks within a project, assigned to users with deadlines and statuses.",
      "constraints": {},
      "indexes": {
        "index_tasks_on_assigned_user_id": {
          "columns": [
            "assigned_user_id",
          ],
          "name": "index_tasks_on_assigned_user_id",
          "type": "",
          "unique": false,
        },
        "index_tasks_on_project_id": {
          "columns": [
            "project_id",
          ],
          "name": "index_tasks_on_project_id",
          "type": "",
          "unique": false,
        },
      },
      "name": "tasks",
    },
    "timesheets": {
      "columns": {
        "duration_minutes": {
          "check": null,
          "comment": "The total duration of work in minutes.",
          "default": null,
          "name": "duration_minutes",
          "notNull": true,
          "primary": false,
          "type": "integer",
          "unique": false,
        },
        "end_time": {
          "check": null,
          "comment": "The timestamp when the user stopped working on the task.",
          "default": null,
          "name": "end_time",
          "notNull": false,
          "primary": false,
          "type": "timestamp",
          "unique": false,
        },
        "id": {
          "check": null,
          "comment": null,
          "default": null,
          "name": "id",
          "notNull": true,
          "primary": true,
          "type": "uuid",
          "unique": true,
        },
        "start_time": {
          "check": null,
          "comment": "The timestamp when the user started working on the task.",
          "default": null,
          "name": "start_time",
          "notNull": true,
          "primary": false,
          "type": "timestamp",
          "unique": false,
        },
        "task_id": {
          "check": null,
          "comment": "Foreign key linking the timesheet to a task.",
          "default": null,
          "name": "task_id",
          "notNull": true,
          "primary": false,
          "type": "uuid",
          "unique": false,
        },
        "user_id": {
          "check": null,
          "comment": "Foreign key linking the timesheet to a user.",
          "default": null,
          "name": "user_id",
          "notNull": true,
          "primary": false,
          "type": "uuid",
          "unique": false,
        },
      },
      "comment": "Tracks time spent by users on tasks for reporting or billing purposes.",
      "constraints": {},
      "indexes": {
        "index_timesheets_on_task_id": {
          "columns": [
            "task_id",
          ],
          "name": "index_timesheets_on_task_id",
          "type": "",
          "unique": false,
        },
        "index_timesheets_on_user_id": {
          "columns": [
            "user_id",
          ],
          "name": "index_timesheets_on_user_id",
          "type": "",
          "unique": false,
        },
      },
      "name": "timesheets",
    },
    "user_roles": {
      "columns": {
        "id": {
          "check": null,
          "comment": null,
          "default": null,
          "name": "id",
          "notNull": true,
          "primary": true,
          "type": "uuid",
          "unique": true,
        },
        "role_id": {
          "check": null,
          "comment": "Foreign key linking to the role.",
          "default": null,
          "name": "role_id",
          "notNull": true,
          "primary": false,
          "type": "uuid",
          "unique": false,
        },
        "user_id": {
          "check": null,
          "comment": "Foreign key linking to the user.",
          "default": null,
          "name": "user_id",
          "notNull": true,
          "primary": false,
          "type": "uuid",
          "unique": false,
        },
      },
      "comment": "Associates users with roles to define their permissions within the company.",
      "constraints": {},
      "indexes": {
        "index_user_roles_on_user_id_and_role_id": {
          "columns": [
            "user_id",
            "role_id",
          ],
          "name": "index_user_roles_on_user_id_and_role_id",
          "type": "",
          "unique": true,
        },
      },
      "name": "user_roles",
    },
    "users": {
      "columns": {
        "age": {
          "check": null,
          "comment": "The user's age.",
          "default": 30,
          "name": "age",
          "notNull": false,
          "primary": false,
          "type": "integer",
          "unique": false,
        },
        "company_id": {
          "check": null,
          "comment": "Foreign key linking the user to a company.",
          "default": null,
          "name": "company_id",
          "notNull": true,
          "primary": false,
          "type": "uuid",
          "unique": false,
        },
        "created_at": {
          "check": null,
          "comment": "The timestamp when the user was created.",
          "default": null,
          "name": "created_at",
          "notNull": true,
          "primary": false,
          "type": "timestamp",
          "unique": false,
        },
        "department_id": {
          "check": null,
          "comment": "Foreign key linking the user to a department.",
          "default": null,
          "name": "department_id",
          "notNull": true,
          "primary": false,
          "type": "uuid",
          "unique": false,
        },
        "id": {
          "check": null,
          "comment": null,
          "default": null,
          "name": "id",
          "notNull": true,
          "primary": true,
          "type": "uuid",
          "unique": true,
        },
        "is_deleted": {
          "check": null,
          "comment": "Indicates whether the user is deleted (soft delete).",
          "default": false,
          "name": "is_deleted",
          "notNull": false,
          "primary": false,
          "type": "boolean",
          "unique": false,
        },
        "name": {
          "check": null,
          "comment": "The user's full name.",
          "default": "new user",
          "name": "name",
          "notNull": false,
          "primary": false,
          "type": "varchar",
          "unique": false,
        },
        "user_name": {
          "check": null,
          "comment": "A unique identifier or login name for the user.",
          "default": null,
          "name": "user_name",
          "notNull": true,
          "primary": false,
          "type": "varchar",
          "unique": false,
        },
      },
      "comment": "Represents employees or members of a company, who are assigned roles and tasks.",
      "constraints": {},
      "indexes": {
        "index_users_on_company_id": {
          "columns": [
            "company_id",
          ],
          "name": "index_users_on_company_id",
          "type": "",
          "unique": false,
        },
        "index_users_on_department_id": {
          "columns": [
            "department_id",
          ],
          "name": "index_users_on_department_id",
          "type": "",
          "unique": false,
        },
      },
      "name": "users",
    },
  },
}
`;
